File: mfcc.m
function c = mfcc(s, fs)
% MFCC Calculate the mel frequencey cepstrum coefficients (MFCC) of a signal
% Inputs:
%       s       : speech signal
%       fs      : sample rate in Hz
% Outputs:
%       c       : MFCC output, each column contains the MFCC's for one speech frame

N = 256;
M=100;
win_frames = zeros(N , floor((length(s)-N+M)/M-1));
frames=zeros(N , floor((length(s)-N+M)/M-1));

n=[0:N-1]';
w_n = 0.54 - 0.46*cos(2*pi*n/(N-1));

for i = 0 : floor((length(s)-N+M)/M-1)
    %the columns of frames are the blocked signal pieces.
    frame = s( i*M+1 : i*M+1 + N -1);
    
%     Can use the below two lines to plot the image of the frame blocked signal
%     Can use the same imagesc to show signal after Hamming windowing (can
%     see the signal tapering off in each signal block)
     frames(:,i+1) = s( i*M+1 : i*M+1 + N -1);
%     imagesc(frames)

    %applys the Hamming windowing to the framed signal;
    win_frames(: , i+1) = frame.*w_n;

end

fft_frames = zeros(size(win_frames));
%for loop computes the fft of the Hamming windowed frames. 
for i = 1 : size(win_frames , 2)
   fft_frames(:,i) = fft(win_frames(:,i)); 
   %imagesc(abs(fft_frames))
end

%Determine matrix for a mel-spaced filterbank
 m = melfb(20, N, fs);
 N2 = 1 + floor(N/2);
Sk=zeros(size(m,1), size(fft_frames,2));
%to plot mel filterbanks:
% plot(linspace(0, (12500/2), 129), melfb(20, 256, 12500)')

%Compute the mel-scale spectrum of the FFT frames
for i = 1 : size(fft_frames,2)
 Sk(:,i) = m * abs(fft_frames(1:N2,i)).^2;
 %to plot mel frequency spectrum: imagesc(Sk)
end

ms = Sk;
c = [];
for i = 1 : size(ms,2)
    %computes the mel-freequency cepstrum coefficients of each column of
    %the mel-frequency spectrum
    c(:, i) = dct(log(ms(:,i)));
end

%excludes the 0'th order cepstral coefficient; each column is the cepstrum
%coefficient for one of the blocked frames.
c = c(2:end , :);

end

File: MFCC.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Mel Frequency Cepstrum Coefficients class

This moves low and high frequencies to different sides 
of the spectrum thus giving us a non-linear output 
for human auditory perception

@author: alejandrogonzales
"""
import numpy as np
from scipy.fftpack import dct

# Mel Frequency Cepstrum Coefficients Class
# TODO remove all refs to self and replace with local vars
class MFCC():

	# Initialize input constructor vars
	Afcc = 1125	# amplitude of MFCC calculation (default see Mel scale equation)	
	#Afcc = 2595 # amplitude of MFCC calculation (default see Mel scale equation)	
	Bfcc = 700	# amplitude for exponential calculation (see Mel scale equation)	
	Nfft = 0	# number of FFT pts	
	Nfcc = 0	# number of MFCC filterbanks	
	Fs = 0		# sampling frequency	
	lower_freq = 0
	upper_freq = 0
	mel_pts = None	# init mel scale array
	freq_pts = None	# init frequency array
	bin_pts = None	# frequency bin array

	# Constructor
	def __init__(self, Nfft, Nfcc, Fs, lower_freq, upper_freq):
		self.Nfft = Nfft	
		#self.Nfcc = Nfcc+2	# add two to account for indep. endpoints	
		self.Nfcc = Nfcc	
		self.Fs = Fs	
		print("Nfft = " + str(Nfft))
		print("Nffc = " + str(Nfcc))
		print("Fs = " + str(Fs))
		print("\n")
		self.lower_freq = lower_freq
		self.upper_freq = upper_freq 

		# Initialize the mel frequency scale
		mel_pts = np.zeros(Nfcc+2)
		freq_pts = np.zeros(Nfcc+2)

	# Fit the mel frequency coefficients
	def fit(self):
		# TODO Consider create a function to calculate all mel coeffs

		# Fit the mel scale, frequency pts and frequency bin pts (filterbank frames) 
		mfccMap = {}	
		mel_lower = self.Afcc*np.log(1 + self.lower_freq/self.Bfcc)
		mel_upper = self.Afcc*np.log(1 + self.upper_freq/self.Bfcc)
        #mel_lower = self.Afcc*np.log10(1 + self.lower_freq/self.Bfcc)
		#mel_upper = self.Afcc*np.log10(1 + self.upper_freq/self.Bfcc)
		self.mel_pts = np.linspace(mel_lower, mel_upper, self.Nfcc+2)	
		self.freq_pts = self.Bfcc*(np.exp(self.mel_pts/self.Afcc) - 1) 
		#self.freq_pts = self.Bfcc*(10**(self.mel_pts/self.Afcc) - 1) 
		self.bin_pts = np.floor((self.Nfft)*self.freq_pts/self.Fs)

		mfccMap['mel'] = self.mel_pts
		mfccMap['freq'] = self.freq_pts
		mfccMap['bin'] = self.bin_pts	
	
		return mfccMap

	# Calculate the filterbanks (i.e. triangular filters)
	def calc_filter_banks(self):
		# initalize filterbank matrix and other dep. vars 
		Nfcc = self.Nfcc
		bin_pts = self.bin_pts

		# Matrix of filterbanks (rowlen = num_filteranks, colen = numfft)
		fbank = np.zeros((Nfcc, int(np.floor((self.Nfft/2+1)))))
		# Loop through the rows of the filterbank matrix (one row per bank)	
		for m in np.arange(1, Nfcc+1):
			f_m_minus = int(bin_pts[m-1])	# left triangle pt
			f_m = int(bin_pts[m])			# center triangle pt
			f_m_plus = int(bin_pts[m+1])	# right triangle pt

			# Loop through left triangular pts (see Hm(k) function)	
			for k in np.arange(f_m_minus, f_m):
				fbank[m-1, k] = (k - bin_pts[m-1])/(bin_pts[m] - bin_pts[m-1])	
			
			# Loop through right triangular pts (see Hm(k) function)
			for k in np.arange(f_m, f_m_plus):
				fbank[m-1, k] = (bin_pts[m+1] - k)/(bin_pts[m+1] - bin_pts[m])
				

		return fbank	

    # Calculate the Mel Frequency Cepstrum Coefficients (ie acoustic vector)
    # Each utterance is transformed into sequence of acoustic vectors
	def calc_mfcc_dct(self, filter_banks, num_mfcc):
		mfcc = dct(filter_banks, type=2, axis=1, norm='ortho')
		mfcc = mfcc[:, 1:(num_mfcc+1)]	# pull the 2nd to the num_mfcc+1 entries	
		
		return mfcc
	
	# Apply sin lifter to the MFCC matrix
	def calc_mfcc_lift(self, mfcc, num_lift):
		# apply liftering to remove high quefrencies from the MFCCs
		[nframes, ncoeffs] = mfcc.shape	# get the coefficients
		n = np.arange(ncoeffs)			# indices for mfcc coefficients
		slift = 1 + (num_lift/2)*np.sin((np.pi*n)/num_lift)	# sin lifter for high quefrencies
		mfcc_lift = mfcc*slift	


		# TODO To improve the MFCCs, that is to reduce the dependency on higher and
		# and lower frequencies in the vocal tract we need to apply liftering, that
		# is filtering in the quefrency domain (ie cepstrum domain)

		return mfcc_lift
